lightstep/sdk/metric/internal/syncstate/sync.go

```diff
@@ -165,55 +169,83 @@ func capture[N number.Any, Traits number.Traits[N]](_ context.Context, inst *Ins
 		return
 	}
 
-	rec, updater := acquireRecord[N](inst, attrs)
+	rec := acquireRecord[N](inst, attrs)
 	defer rec.refMapped.unref()
 
-	updater.Update(num)
+	rec.accumulator.(viewstate.Updater[N]).Update(num)
 
 	// Record was modified.
 	atomic.AddInt64(&rec.updateCount, 1)
 }
 
-// acquireRecord gets or creates a `*record` corresponding to `attrs`,
-// the input attributes.
-func acquireRecord[N number.Any](inst *Instrument, attrs []attribute.KeyValue) (*record, viewstate.Updater[N]) {
-	aset := attribute.NewSet(attrs...)
+// acquireRead acquires the read lock and searches for a `*record`.
+func acquireRead(inst *Instrument, aset attribute.Set) *record {
+	inst.lock.RLock()
+	defer inst.lock.RUnlock()
 
-	if lookup, ok := inst.current.Load(aset); ok {
+	if rec, ok := inst.current[aset]; ok {
 		// Existing record case.
-		rec := lookup.(*record)
-
 		if rec.refMapped.ref() {
 			// At this moment it is guaranteed that the
 			// record is in the map and will not be removed.
-			return rec, rec.accumulator.(viewstate.Updater[N])
+			return rec
 		}
-		// This record is no longer mapped, try to add a new
-		// record below.
 	}
+	return nil
+}
+
+// acquireRecord gets or creates a `*record` corresponding to `attrs`,
+// the input attributes.
+func acquireRecord[N number.Any](inst *Instrument, attrs []attribute.KeyValue) *record {
+	aset := attribute.NewSet(attrs...)
+
+	rec := acquireRead(inst, aset)
+	if rec != nil {
+		return rec
+	}
+
+	// Record is not mapped.
 
 	// Note: the accumulator set below is created speculatively;
-	// if it is never returned, it will be released below.
+	// it will be released if it is never returned.
 	newRec := &record{
 		refMapped:   newRefcountMapped(),
 		accumulator: inst.compiled.NewAccumulator(aset),
 	}
 
 	for {
-		if found, loaded := inst.current.LoadOrStore(aset, newRec); loaded {
-			oldRec := found.(*record)
-			if oldRec.refMapped.ref() {
-				newRec.accumulator.SnapshotAndProcess(true)
-				return oldRec, oldRec.accumulator.(viewstate.Updater[N])
-			}
+		acquired, repeat := acquireWrite(inst, aset, newRec)
+
+		if repeat {
 			// When this happens, we are waiting for the call to Delete()
 			// inside SnapshotAndProcess() to complete before inserting
 			// a new record.  This avoids busy-waiting.
 			runtime.Gosched()
 			continue
 		}
-		break
+
+		if acquired != newRec {
+			// Release the speculative accumulator, since it was
+			newRec.accumulator.SnapshotAndProcess(true)
+		}
+		return acquired
+	}
+}
+
+// acquireWrite acquires the write lock and gets or sets a `*record`.
+func acquireWrite(inst *Instrument, aset attribute.Set, newRec *record) (*record, bool) {
+	inst.lock.Lock()
+	defer inst.lock.Unlock()
+
+	oldRec, loaded := inst.current[aset]
+
+	if loaded {
+		if oldRec.refMapped.ref() {
+			return oldRec, false
```

Comment:
nit: Typically, I expect this type of "get" function to return a boolean that indicates if the first return value is valid or not, which is the opposite here.

I can understand it if I read `acquired, repeat := acquireWrite(inst, aset, newRec)` first - but it can be confusing for someone looking at this method definition without seeing its usage.